### PR TITLE
Add developer documentation on e2e testing.

### PR DESCRIPTION
**Pull request by timothysc**
Add developer documentation for e2e testing, such that Q/E and other folks can leverage. 

/cc @eparis @pmorie @jayunit100

### CODE REVIEW

**Comment by eparis** [comment]
@ncdc @smarterclayton

File: docs/devel/e2e-tests.md [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+## Performance Evaluation
+
+Another benefit of the end-2-end tests is the ability to create reproducible loads on the system, which can then be used to determine the responsiveness, or analyze other characteristics of the system.  For example, the density tests load the system to 30,50,100 pods per/node and measures the different characteristics of the system, such as throughput, api-latency, etc.
+
+For a good overview of how we analyze performance data, please read the following [post](http://blog.kubernetes.io/2015/09/kubernetes-performance-measurements-and.html)
+
+Developers who are interested in doing their own performance analysis, we recommend setting up [prometheus](http://prometheus.io/) for data collection, and using [promdash](http://prometheus.io/docs/visualization/promdash/) to visualize the data.  There also exists the option of pushing your own metrics in from the tests using a [prom-push-gateway](http://prometheus.io/docs/instrumenting/pushing/).  Containers for all of these components can be found [here](https://hub.docker.com/u/prom/).
```
Comment (jayunit100) [inline docs/devel/e2e-tests.md]: s/Develeopers/For developers

File: docs/devel/e2e-tests.md [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+Prior to running the tests, it is recommended that you first create a simple auth file in your home directory, e.g. `$HOME/.kubernetes_auth` , with the following:
+
+```
+{
+  "User": "root",
+  "Password": ""
+}
+```
+
+Next, you will need a cluster that you can test against.  As mentioned earlier, you will want to execute `sudo ./hack/local-up-cluster.sh`.  To get a sense of what tests exist, you may want to run:
```
Comment (jayunit100) [inline docs/devel/e2e-tests.md]: Might want to mention "If you're doing alot of development with local-up-cluster, you may want to clear your genereated credentials and refresh iptables periodically."

- `rm -rf /var/run/kubernetes` (clear kube generated credentials, sometimes stale ones/bad perms can cause problems.
- `sudo iptables -F` (clear ip tables rules, also if doing alot of dev and running local e2e's, this is necessary every so often).

**Comment by jayunit100** [comment]
Good tutorial, some minor ideas/comments, also pinged jon cope on it offline, who is a typical audience for this doc.  Overall LGTM to me and good to merge as first iteration.

**Comment by timothysc** [comment]
I'll wait for other comments, then change in bulk.

File: docs/devel/e2e-tests.md (hunk 2 above)
Comment (timothysc) [inline docs/devel/e2e-tests.md]: Done.

**Comment by pmorie** [comment]
@timothysc You should add a link to this from the api-changes doc; there's text about needing to write this doc in that one.

**Comment by timothysc** [comment]
@pmorie happy to do a follow on PR for links + cleanup.

**Comment by timothysc** [comment]
@eparis Either the bot is super slow, or label magic has shifted yet again ... ?

**Comment by nikhiljindal** [comment]
I just tried this and I am getting the following error:
```
--- FAIL: TestE2E (0.00s)
	e2e_test.go:132: Error waiting for all pods to be running and ready: error creating client: KubeConfig must be specified to load client config
FAIL
```

The command I ran is:
```
./_output/local/bin/linux/amd64/e2e.test --host="127.0.0.1:8080" --provider="local" --ginkgo.v=true -ginkgo.dryRun=true
```
I have a local cluster up which is fine.
Looking at the code, it seems like SetTextContext (https://github.com/kubernetes/kubernetes/blob/master/test/e2e/util.go#L130) is never called, which is why kubeconfig is empty leading to that error: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/util.go#L824

What am I missing?

**Comment by jayunit100** [comment]
ah ! provide the kubeconfig file, it can be empty :) 
```
echo "{}" >> /tmp/dud
--kubeconfig=/tmp/dud
```

**Comment by timothysc** [comment]
@nikhiljindal you never specified your --kubeconfig which was denoted in the docs.

**Comment by ihmccreery** [comment]
(cc @ixdy @quinton-hoole)

Just saw this PR; thanks for these docs @timothysc!

It seems, though, to contradict some of the information in [development.md](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/development.md#end-to-end-tests).  It also doesn't look like it accurately reflects [how we run tests in Jenkins](https://github.com/kubernetes/kubernetes/blob/master/hack/jenkins/e2e.sh).  Was there discussion about removing, linking, or changing existing documentation, and how to keep this from drifting with what we do in Jenkins?

**Comment by timothysc** [comment]
@ihmccreery This is developer oriented so it will certainly be different then the details re: Jenkins.  I did that on purpose b/c upstreams scripts etc are prescriptive, while the details here are general to developer running locally.  

We also automate our testing very differently as well.  At some point we should unify that process, but imho separate doc.

**Comment by ihmccreery** [comment]
Cool, thanks.  Any thoughts on combining or removing content in [development.md](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/development.md#end-to-end-tests)?

**Comment by timothysc** [comment]
@ihmccreery I think it's fine to keep the content, but just add a link to this one.

**Comment by ixdy** [comment]
Yeah, I missed this PR when it went by.

There is some important setup that's done by `hack/ginkgo-e2e.sh`; developers should be calling that (or one of the wrappers which calls it) rather than the `e2e.test` binary directly.

Eventually I'd like to fold much of the functionality from `hack/ginkgo-e2e.sh` into `test/e2e/e2e_test.go`, so we don't need this wrapper, but as-is this doc is incorrect.

**Comment by ixdy** [comment]
... all that said, I really appreciate you writing documentation for this process!

I'll try to take a look tomorrow at this doc and give specific feedback on things where our process differs a bit with yours.

**Pull request merged**